### PR TITLE
Runner.go Context Window Shifting

### PR DESCRIPTION
**Pull request by jessegross**
This series implements context window shifting for the new go server runner. It also fixes a number of issues in the related code.

My intention is to start adding tests for some of the issues encountered here but I wanted to start getting reviews on this code in the meantime.

### CODE REVIEW

File: llama/llama.go [hunk 1]
```diff
@@ -209,11 +223,15 @@ func (m *Model) ApplyLoraFromFile(loraPath string, scale float32, baseModelPath
 }
 
 type Batch struct {
-	c C.struct_llama_batch
+	c         C.struct_llama_batch
+	allocSize int
```
[the remark this answers is not in the document]
Comment (jessegross) [reply llama/llama.go:227]: Sure

File: llama/llama.go [hunk 2]
```diff
@@ -357,6 +390,7 @@ type SamplingParams struct {
 	TfsZ           float32
 	TypicalP       float32
 	Temp           float32
+	RepeatLastN    int
```
Comment (jmorganca) [inline llama/llama.go:393]: Note for later: I believe we're missing `min_p` as well

File: llama/runner/runner.go [hunk 3]
```diff
@@ -116,6 +131,9 @@ type Server struct {
 	// context window size
 	numCtx int
 
+	// does this model require a beginning of sequence token?
+	bosToken int
```
Comment (jmorganca) [inline llama/runner/runner.go:135]: should this be a `bool`?

File: llama/runner/runner.go [hunk 4]
```diff
@@ -288,13 +365,35 @@ func (s *Server) run(ctx context.Context) {
 	}
 }
 
+type Options struct {
+	api.Runner
+
+	NumKeep          int      `json:"n_keep"`
```
Comment (jmorganca) [inline llama/runner/runner.go:371]: Note a huge deal, but why re-declare these?

**Review by jmorganca — APPROVED** [review]

**Comment by jmorganca** [comment]
LGTM minus some small comments + linter error

File: llama/llama.go (hunk 2 above)
Comment (jessegross) [reply llama/llama.go:393]: Good catch, I’ll put that on the list for follow up.

File: llama/runner/runner.go (hunk 3 above)
Comment (jessegross) [reply llama/runner/runner.go:135]: Logically yes, though it gets added to the number of tokens, which is why I made it an int.

File: llama/runner/runner.go (hunk 4 above)
Comment (jessegross) [reply llama/runner/runner.go:371]: The JSON field names are different for communicating to the runner backend (as defined by server.cpp) vs. the Ollama API. I think for now, it is best to just match server.cpp to minimize differences in Ollama while we transition. Once we fully transition we can avoid multiple declarations.

This seemed like the cleanest way to use different JSON fields - is there a better way?

**Comment by jessegross** [comment]
Rebased (which fixes the lint issue) and tacked on a fix for MinP.

File: llama/runner/runner.go (hunk 4 above)
Comment (jmorganca) [reply llama/runner/runner.go:371]: Ah that's right. This makes sense then

**Pull request merged**